api/python/cellxgene_census/src/cellxgene_census/experimental/ml/pytorch.py

```diff
@@ -266,19 +315,26 @@ def __init__(
         self.return_sparse_X = return_sparse_X
         self.encoders = encoders
         self.stats = stats
+        self.gc_elapsed = 0
         self.max_process_mem_usage_bytes = 0
         self.X_dtype = X.schema[2].type.to_pandas_dtype()
 
     def __next__(self) -> ObsAndXDatum:
         """Read the next torch batch, possibly across multiple soma chunks."""
         obs: pd.DataFrame = pd.DataFrame()
-        X: sparse.csr_matrix = sparse.csr_matrix((0, len(self.var_joinids)), dtype=self.X_dtype)
+        X: ChunkX = csr_matrix((0, len(self.var_joinids)), dtype=self.X_dtype)
+        first = True
 
         while len(obs) < self.batch_size:
             try:
                 obs_partial, X_partial = self._read_partial_torch_batch(self.batch_size - len(obs))
-                obs = pd.concat([obs, obs_partial], axis=0)
-                X = sparse.vstack([X, X_partial])
+                if first:
+                    obs = obs_partial
+                    X = X_partial
+                    first = False
+                else:
+                    obs = pd.concat([obs, obs_partial], axis=0)
+                    X = sparse.vstack([X, X_partial])
```

Comment:
Actually, this code is now bugged. `sparse.vstack` doesn't like it when you pass numpy arrays, but this code can now receive numpy arrays since densification now happens in the earlier iterator.

Test in 8bab49a, fix in 4c042ee

Reply:
Thanks @ivirshup, your version is definitely better, just wanted to note/confirm that I'd fixed the vstack issue on a different branch ([@alb]; [relevant diff lines](https://github.com/chanzuckerberg/cellxgene-census/compare/main...ryan-williams:cellxgene-census:alb#diff-229c84bdbd7ccbabfc2e3bdadd52acfdabc53a63e76c20b2f5e4854e35ea0b36R330-R333)), which I ran most recent benchmarks from.

I inadvertently pushed my most recent work to [@alb], though this PR was from [@rw/alb](https://github.com/ryan-williams/cellxgene-census/tree/rw/alb), sorry about that. [@alb] also contains other functionality I might make a new PR for.

[@alb]: https://github.com/ryan-williams/cellxgene-census/tree/alb